src/run_program.rs

```diff
@@ -66,6 +88,7 @@ struct RunProgramContext<'a, D> {
     val_stack: Vec<NodePtr>,
     env_stack: Vec<NodePtr>,
     op_stack: Vec<Operation>,
+    softfork_stack: Vec<SoftforkGuard>,
```

Comment:
As per my comment above, why not just use a stack of dialects, and at the end of the softfork eval process, pop off the top dialect.